src/utils/instrumentation.ts

```diff
@@ -0,0 +1,201 @@
+/*!
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import arrify = require('arrify');
+import path = require('path');
+import {google} from '../../protos/protos';
+import {Entry} from '../entry';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+declare const global: {[index: string]: any};
+
+// The global variable keeping track if instrumentation record was already written or not.
+// The instrumentation record should be generated only once per process and contain logging
+// libraries related info.
+global.instrumentationAdded = false;
+
+// The variable to hold cached library version
+let libraryVersion: string;
+
+// Max length for instrumentation library name and version values
+const maxDiagnosticValueLen = 14;
+
+export const DIAGNOSTIC_INFO_KEY = 'logging.googleapis.com/diagnostic';
+export const INSTRUMENTATION_SOURCE_KEY = 'instrumentation_source';
+export const NODEJS_LIBRARY_NAME_PREFIX = 'nodejs';
```

Comment:
nit: is it approved language code? can we add a link to the list of code references (e.g. in gax/grpc)

Reply:
It is - I used a reference from [OTel](https://github.com/open-telemetry/opentelemetry-specification/tree/a6cfbc92d36aa019c95fc2d9f023f5de5d570937/specification/resource/semantic_conventions#telemetry-sdk) telemetry for this